### PR TITLE
[ci] Speed up checking out the code

### PR DESCRIPTION
**Pull request by tsapeta**
# Why

`Checkout code` step in CI takes definitely too much time and is kinda unpredictable - the longest waiting time I encountered was almost 9 minutes! Unfortunately I cannot find it right now 😅

## Some stats from before:

Screenshot 2019-12-04 at 11 28 32
Screenshot 2019-12-04 at 11 33 14
Screenshot 2019-12-04 at 11 34 42
Screenshot 2019-12-04 at 11 35 21

# How

`git clone` operation can be faster by doing a shallow copy - then we download only the last commit and fetch a few previous ones if we need to compare them with master (to check whether we can halt a job if nothing has changed).

## After these changes:

Screenshot 2019-12-04 at 11 38 27
Screenshot 2019-12-04 at 12 14 15

# Test Plan

All CI jobs in this PR pass.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -122,6 +130,66 @@ commands:
             else
               echo 'user-halt' > "$CIRCLE_INTERNAL_SCRATCH/HALT_TASK"
             fi
+  checkout-shallow:
+    steps:
+      - run:
+          name: Checkout code (shallow)
+          command: |
```
Comment (dsokal) [inline .circleci/config.yml]: Maybe add a comment that you copy-pasted the CircleCI code and slightly modified it :)

File: .circleci/config.yml [hunk 2]
```diff
@@ -272,6 +340,8 @@ executors:
     shell: /run/current-system/sw/bin/bash -leo pipefail
     working_directory: ~/expo
     environment:
+      # Set project directory as environment variable
+      PROJECT_DIRECTORY: /root/expo
```
Comment (dsokal) [inline .circleci/config.yml]: I think you don't have to define your own env var. Have you tried to use `CIRCLE_WORKING_DIRECTORY` instead?
https://circleci.com/docs/2.0/env-vars/#built-in-environment-variables

File: .circleci/config.yml [hunk 3]
```diff
@@ -37,7 +37,11 @@ commands:
               exit 0
             fi
 
-            BASE_COMMIT=$(git merge-base HEAD remotes/origin/"${ROOT_BRANCH}")
+            BASE_COMMIT=$(git merge-base $CIRCLE_BRANCH $ROOT_BRANCH)
+            while [ -z $BASE_COMMIT ]; do
```
Comment (dsokal) [inline .circleci/config.yml]: It's an infinite loop.

File: .circleci/config.yml [hunk 4]
```diff
@@ -122,6 +136,67 @@ commands:
...
               echo 'user-halt' > "$CIRCLE_INTERNAL_SCRATCH/HALT_TASK"
             fi
+  checkout-shallow:
+    steps:
+      - run:
+          name: Checkout code (shallow)
+          command: |
+            # This command has been mostly copied from the `checkout` command available in CircleCI.
+
+            # Workaround old docker images with incorrect $HOME
```
Comment (ide) [inline .circleci/config.yml]: It looks like this was fixed about five years ago -- probably don't need it?

**Review by ide — COMMENTED** [review]
Seems useful to try this. If it works well, let's move it into a CircleCI orb (https://github.com/expo/circleci-orbs) so that we can use it in Universe and Expo CLI as well.

@nicknovitski Could you take a look at this and also check that it meets our style conventions?

**Comment by tsapeta** [comment]
@ide Sadly, I need to look at this more closely as something weird is now happening during `conditionally_halt` and `halt_if_unchanged` commands. I've tried running the same commands on SSH and it worked as expected 🤔

**Comment by nicknovitski** [comment]
I recently touched the custom checkout step in universe.  Since before I've joined, that step has had `--depth=100`.  All I did was move it to an actual custom action, update the script to be identical to the current circleci step, and check how long the step takes with no depth flag, a depth of 100, and a depth of 50.  What I found was that no depth flag was significantly slower, but 50 wasn't noticeably faster than 100.  It makes sense: github's servers had to do more work to create these fake "flattened" commits.

I'd be suprised if a depth of 1 was significantly faster than one of 20 or 30, which might work much better with the build stopping steps.

**Comment by tsapeta** [comment]
Ok, it works as expected now 😉 The problem was with `git merge-base` command that was exiting with code `1` when there is no merge base between given commits. I've muted this command by adding `|| :` at the end.

@ide I've published that command to CircleCI Orb registry, the current code of it is here: https://github.com/expo/circleci-orbs/commit/57ce3e3542889638ad1b26c650cc74aaad008c4a

PR is again ready to review and merge.

**Review by ide — APPROVED** [review]
This looks ok to me. @nicknovitski could you do a post-commit review of https://github.com/expo/circleci-orbs/commit/57ce3e3542889638ad1b26c650cc74aaad008c4a and make whatever changes you think we should make?

**Pull request merged**